### ISSUE TITLE
Progressive Growing of GANs for Improved Quality, Stability and Variation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#18]
![image](https://user-images.githubusercontent.com/46308407/104118209-a54f2b80-536a-11eb-9502-c672537acba4.png)

# Key Idea

- ***Progressively enlarge*** the capacity of Generator/Discriminator to produce ***high resolution image***
- ***Increase Variation*** using ***mini batch standard deviation*** as feature map
- ***Equalize learning rates*** for all weights
- ***Pixelwise feature vector normalization*** in Generator
- Propose ***Sliced Wasserstein Distance*** as new metric evaluating both variation and quality

# Previous Approaches

- Hierarchical GANs : define generators and discriminators for each level of an image pyramid
- Used multiple discriminators, generators or both for different spatial resolution
- Major difference lies in that ProGAN uses single GAN, while previous approaches employed multiple networks for each resolution level

# Formulation

![image](https://user-images.githubusercontent.com/46308407/104118231-c6b01780-536a-11eb-898a-17969726c089.png)

- All layers remain trainable once introduced.
- Gradually sneak in layers for larger resolutions using ***fade in*** technique.

![image](https://user-images.githubusercontent.com/46308407/104118236-d0397f80-536a-11eb-87aa-fe09c1807336.png)

- $\alpha$ linearly increased from 0 to 1, to ***avoid sudden shock to well trained lower-resolution layers.***

### Mini Batch Standard Deviation

***[GOAL] Make level of variance similar to batch of real images or else discriminator will find out***

- Obtain standard deviation across feature maps for each pixel
- Average the standard deviations to get a single value
- Replicate the single value to the desired dimension ($H*W$) to form additional feature map
- Concat the feature map channel-wise towards the end in the discriminator network.

### Equalize Learning Rates

![image](https://user-images.githubusercontent.com/46308407/104118251-eb0bf400-536a-11eb-9edc-b20216614776.png)

- Above is formulation of Adam Optimizer
- Each weight uses its squared sum of past gradients (sd-like component)
- Learning rate for each weight differs, so a universal lr can be too high and low at the same time

### Pixel-wise Feature Vector Normalization

- Major problem in GANs is that the signal magnitude spirals out of control (exploding gradient)
![image](https://user-images.githubusercontent.com/46308407/104118263-ff4ff100-536a-11eb-88d1-79af10f562a4.png)

- Simply normalize the feature vector for each pixel so that $||l||_k=1$
- This heavy-handed constraint does NOT harm the generator but prevent rise of signal size

### Sliced Wasserstein Distance [SWD]

- Multi-scale statistical similarity between distributions of local image patches

# Experiments
![image](https://user-images.githubusercontent.com/46308407/104118268-0d057680-536b-11eb-8ba8-e28fd7204425.png)

- Ablation Study for each element they propose.
- Note that (c) Small mini-batch is due to GPU memory when trying to generate $1024^2$ high resolution images

![image](https://user-images.githubusercontent.com/46308407/104118283-1a226580-536b-11eb-81b0-4d85537b046c.png)

- Note that (b) is with progressive growing. Each sequentially converges, and does so faster and beetter. ($1/2$ training time)

# Comments

- It's very intuitive that to generate high resolution image, one has to do so step by step.
- Similar with human painting. First sketch the oval face, than fill in eyes, nose, mouth, and then add fine features like wrinkles and color.
- The fade in technique is interesting. Also quite intuitive and clever.
- Surprised at how pixel-wise feature vector normalization doesn't harm the generation. It looks to much of a constraint to me, as it can change relative values of pixels in a given feature map.
